docs/codeql/codeql-language-guides/customizing-library-models-for-ruby.rst

```diff
@@ -0,0 +1,406 @@
+.. _customizing-library-models-for-ruby:
+
+:orphan:
+:nosearch:
+
+Customizing Library Models for Ruby
+===================================
+
+.. include:: ../reusables/beta-note-customizing-library-models.rst
+
+Ruby analysis can be customized by adding library models in data extension files.
+
+A data extension for Ruby is a YAML file of the form:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/ruby-all
+        extensible: <name of extensible predicate>
+      data:
+        - <tuple1>
+        - <tuple2>
+        - ...
+
+The CodeQL library for JavaScript exposes the following extensible predicates:
+
+- **sourceModel**\(type, path, kind)
+- **sinkModel**\(type, path, kind)
+- **typeModel**\(type1, type2, path)
+- **summaryModel**\(type, path, input, output, kind)
+
+See the `CLI documentation for how to load and use data extensions in a CodeQL evaluation run <https://docs.google.com/document/d/14IYCHX8wWuU-HTvJ2gPSdXQKHKYbWCHQKOgn8oLaa80/edit#heading=h.m0v53lpi6w2n>`__ (internal access required).
+
+We'll explain how to use these using a few examples, and provide some reference material at the end of this article.
+
+Example: Taint sink in the 'tty-command' gem
+--------------------------------------------
+
+In this example, we'll show how to add the following argument, passed to **tty-command**, as a command-line injection sink:
+
+.. code-block:: ruby
+
+  tty = TTY::Command.new
+  tty.run(cmd) # <-- add 'cmd' as a taint sink
+
+For this example, you can use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/ruby-all
+        extensible: sinkModel
+      data:
+        - ["TTY::Command", "Method[run].Argument[0]", "command-injection"]
+
+
+- Since we're adding a new sink, we add a tuple to the **sinkModel** extensible predicate.
+- The first column, **"TTY::Command"**, identifies a set of values from which to begin the search for the sink.
+  The string **"TTY::Command""** means we start at the places where the codebase constructs instances of the class **TTY::Command**.
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+
+  - **Method[run]** selects calls to the **run** method of the **TTY::Command** class.
+  - **Argument[0]** selects the first argument to calls to that member.
+
+- **command-injection** indicates that this is considered a sink for the command injection query.
+
+Example: Taint sources from `sinatra` block parameters
+------------------------------------------------------
+
+In this example, we'll show how the `x` parameter below could be marked as a remote flow source:
```

Comment:
```suggestion
In this example, we'll show how the 'x' parameter below could be marked as a remote flow source:
```

Another one where I think it'll otherwise render in italics, and would be ideal to use the same format throughout the doc